### ISSUE TITLE
Less or no suggestion for exclamation mark instead of false comparison

### ISSUE THREAD

**Issue by ScreamingDev** [kalessil/phpinspectionsea#1181]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | False-Positive |
| Plugin         | Php Inspections (EA Extended) or Php Inspections (EA Ultimate) |
| Language level |                |

Current behaviour (description/screenshot:)
--

```
if ( false === $migration instanceof MigrationInterface ) {
```

Says

> '!$migration instanceof MigrationInterface' would fit better here (reduces cognitive load). more... (Strg+F1)

Expected behaviour
--

No warning.

Notes
---

I disagree with this warning for this scenario:

```
if ( ! $migration instanceof MigrationInterface )
```

While on a bug hunt a new colleague asked me what comes first: The `! $migration` or the `instanceof`. We changes it to `false === ...` so you have less cognitive load.

But **even more** I disagree with such hint if you show this warning for any type of value. For example:

```
$i = false;
$i = null;
$i = 0;
$i = 5;

if ( ! $i ) {
  echo 'douh!';
}
```

There is a code analysis tool for PHP that checks for such expressions where the dev was "too lazy" to have proper types. I agree with that tool. Being specific about the things that we cover/check is a good thing and leads to less side effects.

So I hope this hint about `!` is simpler than `false ===` does take care of it's context or does only appear when it is absolute sure that this does not make a strict statement more loose / inject side effects. I know I can disable that but I also not want the good PHPInspections EA to lead rookie developer into false-positives. (Just a hint, no offense)

**Comment by andreasschroth** [comment]
I disagree. For me, `===` and `instanceof` are both operators and in your above example the cognitive load is definitely higher than it would be with `!`, especially without parenthesis, because you have to think about in which order the comparison is executed.

`if ( false === ($migration instanceof MigrationInterface) ) {` would have a better readability imo.

But `if ( ! $migration instanceof MigrationInterface) {` being by far the easiest to read version.

Why not just turn off the inspection?

**Comment by kalessil** [comment]
The inspections is getting lots of negative feedback, so it'll be disabled by default to not distract folks during bug-hunts =_

**Comment by andreasschroth** [comment]
@kalessil Yeah sounds like a good idea to disable it by default. But personally I like it & the suggestions it provides. But yes, it's definitely a subjective thing.

**Comment by kalessil** [comment]
@ScreamingDev: no worries, I appreciate the feedback and accepting the point (cognitive load). We probably mean different thing behind "negative feedback", in this project context it means that multiple people didn't find the inspection useful and disabling it in their projects.

Good point that the feedback brings major drawbacks (cognitive load, conflicts with other inspections, violating some specific code styles and more), so decision to disable the inspection by default was pretty easy to make =)

**Comment by kalessil** [comment]
The inspection is disabled by default.
